subsys/net/lib/lwm2m/lwm2m_obj_firmware_pull.c

```diff
@@ -120,111 +223,249 @@ do_firmware_transfer_reply_cb(const struct zoap_packet *response,
 	}
 
 	return ret;
 }
 
-static enum zoap_block_size default_block_size(void)
+static void do_transmit_timeout_cb(struct lwm2m_message *msg)
 {
-	switch (CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_COAP_BLOCK_SIZE) {
-	case 16:
-		return ZOAP_BLOCK_16;
-	case 32:
-		return ZOAP_BLOCK_32;
-	case 64:
-		return ZOAP_BLOCK_64;
-	case 128:
-		return ZOAP_BLOCK_128;
-	case 256:
-		return ZOAP_BLOCK_256;
-	case 512:
-		return ZOAP_BLOCK_512;
-	case 1024:
-		return ZOAP_BLOCK_1024;
-	}
+	const u8_t *token;
+	u8_t tkl;
 
-	return ZOAP_BLOCK_256;
+	if (firmware_retry < PACKET_TRANSFER_RETRY_MAX) {
+		/* retry block */
+		SYS_LOG_WRN("TIMEOUT - Sending a retry packet!");
+		token = zoap_header_get_token(&msg->zpkt, &tkl);
+
+		transfer_request(&firmware_block_ctx, token, tkl,
+				 do_firmware_transfer_reply_cb);
+		firmware_retry++;
+	} else {
+		SYS_LOG_ERR("TIMEOUT - Too many retry packet attempts! "
+			    "Aborting firmware download.");
+		lwm2m_firmware_set_update_result(RESULT_CONNECTION_LOST);
+	}
 }
 
 static void firmware_transfer(struct k_work *work)
 {
-	int ret, port, family;
+	int ret, family;
+	u16_t off;
+	u16_t len;
+	char tmp;
+	char *server_addr;
 
 	/* Server Peer IP information */
-	/* TODO: use parser on firmware_uri to determine IP version + port */
-	/* TODO: hard code IPv4 + port for now */
-	port = 5685;
 	family = AF_INET;
 
-#if defined(CONFIG_NET_IPV6)
-	if (family == AF_INET6) {
+#if defined(CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_COAP_PROXY_SUPPORT)
+	server_addr = CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_COAP_PROXY_ADDR;
+	if (strlen(server_addr) >= PROXY_URI_LEN) {
+		SYS_LOG_ERR("Invalid Proxy URI: %s", server_addr);
+		lwm2m_firmware_set_update_result(RESULT_UNSUP_PROTO);
+		return;
+	}
+
+	/* Copy required as it gets modified when port is available */
+	strcpy(proxy_uri, server_addr);
+	server_addr = proxy_uri;
+#else
+	server_addr = firmware_uri;
+#endif
+
+	http_parser_url_init(&parsed_uri);
+	ret = http_parser_parse_url(server_addr,
+				    strlen(server_addr),
+				    0,
+				    &parsed_uri);
+	if (ret != 0) {
+		SYS_LOG_ERR("Invalid firmware URI: %s", server_addr);
+		lwm2m_firmware_set_update_result(RESULT_INVALID_URI);
+		return;
+	}
+
+	/* Check schema and only support coap for now */
+	if (!(parsed_uri.field_set & (1 << UF_SCHEMA))) {
+		SYS_LOG_ERR("No schema in package uri");
+		lwm2m_firmware_set_update_result(RESULT_INVALID_URI);
+		return;
+	}
+
+	/* TODO: enable coaps when DTLS is ready */
+	off = parsed_uri.field_data[UF_SCHEMA].off;
+	len = parsed_uri.field_data[UF_SCHEMA].len;
+	if (len != 4 || memcmp(server_addr + off, "coap", 4)) {
+		SYS_LOG_ERR("Unsupported schema");
+		lwm2m_firmware_set_update_result(RESULT_UNSUP_PROTO);
+		return;
+	}
+
+	if (!(parsed_uri.field_set & (1 << UF_PORT))) {
+		/* Set to default port of CoAP */
+		parsed_uri.port = 5683;
+	}
+
+	off = parsed_uri.field_data[UF_HOST].off;
+	len = parsed_uri.field_data[UF_HOST].len;
+
+	/* IPv6 is wrapped by brackets */
```

Comment:
Line 398 ~ 447 can be removed and instead we can pass host URI string to net_app_init_udp_client()